### PR TITLE
DOC/TST: Update the parquet (pyarrow >= 0.15) docs and tests regarding Categorical support

### LINKED ISSUE #27955: DOC/TST: update the parquet (pyarrow >= 0.15) docs and tests regarding Categorical support
Wes is doing great work in Apache Arrow on parquet's categorical support, which means that roundtripping to parquet with `to_parquet`/`read_parquet` will preserve categorical dtypes (and with a much better performance as before). 

See https://issues.apache.org/jira/browse/ARROW-3246 (and linked issues), https://github.com/apache/arrow/pull/5110

We will need to:

- update the tests for pyarrow to test this faithful roundtrip (depending on the pyarrow version): https://github.com/pandas-dev/pandas/blob/802f67046bbae0a815b2fe9d20d2217485bbc942/pandas/tests/io/test_parquet.py#L409, https://github.com/pandas-dev/pandas/blob/802f67046bbae0a815b2fe9d20d2217485bbc942/pandas/tests/io/test_parquet.py#L451
- update the documentation. Eg the caveats section at https://dev.pandas.io/user_guide/io.html#parquet

### PR DESCRIPTION
**Pull request by galuhsahid**
- [x] closes #27955
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by galuhsahid** [comment]
Several questions/thoughts:
1. Should I also add a test in this part: https://github.com/pandas-dev/pandas/blob/802f67046bbae0a815b2fe9d20d2217485bbc942/pandas/tests/io/test_parquet.py#L409 as well? I thought adding a test there would be a duplicate of what I'm doing in https://github.com/pandas-dev/pandas/blob/802f67046bbae0a815b2fe9d20d2217485bbc942/pandas/tests/io/test_parquet.py#L451, but maybe I'm missing something
2. In https://github.com/apache/arrow/pull/5110 the test in pyarrow demonstrates how category values will be preserved regardless whether they occur in the data or not; should the test in pandas demonstrate this as well?

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks a lot for the PR! 

Looks very good. One suggestion I have is to add a categorical column in the first example in the docs (the `df` with columns of all different types that gets created a bit below the place you edited in io.rst).

**Comment by jorisvandenbossche** [comment]
> Should I also add a test in this part: 

I think your analysis is correct that it would be duplicative. 
I think once we don't have the version check anymore (in the future if we would only support pyarrow >= 0.15), then we can move the ``test_categorical`` into the `test_basic`. But for now I would keep it as is.

> In apache/arrow#5110 the test in pyarrow demonstrates how category values will be preserved regardless whether they occur in the data or not; should the test in pandas demonstrate this as well?

Yes, I think that is nice to test. In addition, you could also test if the `ordered` flag is preserved as well. You can add additional columns to the `df` with different cases in `test_categorical`.

**Comment by galuhsahid** [comment]
Thanks for the review! I've added another test for null, out-of-order values & unobserved category and updated the doc as well.

However when I tried to add another test to check if order is preserved, it seems like it's not preserved. I'm not really sure if this is expected or if I'm missing something:

```
import pandas as pd
from pandas.io.parquet import read_parquet, to_parquet

df = pd.DataFrame()
df["a"] = pd.Categorical(["a", "b", "c", "a"], categories=["b", "c", "d"], ordered=True)

df.to_parquet(<path>)
actual = read_parquet(<path>)

df["a"]    
0    NaN
1      b
2      c
3    NaN
Name: a, dtype: category
Categories (3, object): [b < c < d]

actual["a"]
0    NaN
1      b
2      c
3    NaN
Name: a, dtype: category
Categories (3, object): [b, c, d]

```

If it's expected then maybe I can add another test that expects it to still return a categorical data type without order, if needed.

**Comment by jorisvandenbossche** [comment]
> However when I tried to add another test to check if order is preserved, it seems like it's not preserved. I'm not really sure if this is expected or if I'm missing something:

Indeed, I see the same trying your code. That is certainly not expected I think. Can you open an issue for this? (Arrow uses a JIRA tracker: https://issues.apache.org/jira/projects/ARROW/issues; I can also open an issue if you want)

**Review by jorisvandenbossche — APPROVED** [review]
The additional changes look good to me, thanks!

**Comment by jorisvandenbossche** [comment]
Hmm, the failure on Azure is actually related:

```=================================== FAILURES ===================================
_____________________ TestParquetPyArrow.test_categorical ______________________
[gw0] darwin -- Python 3.5.6 /Users/vsts/miniconda3/envs/pandas-dev/bin/python

self = <pandas.tests.io.test_parquet.TestParquetPyArrow object at 0x146354a90>
pa = 'pyarrow'

    def test_categorical(self, pa):
    
        # supported in >= 0.7.0
        df = pd.DataFrame()
        df["a"] = pd.Categorical(list("abcdef"))
    
        # test for null, out-of-order values, and unobserved category
        dtype = pd.CategoricalDtype(["foo", "bar", "baz"])
        df["b"] = pd.Categorical.from_codes(codes=[1, 0, 0, 1, -1, 1], dtype=dtype)
    
        if LooseVersion(pyarrow.__version__) >= LooseVersion("0.15.0"):
            check_round_trip(df, pa)
        else:
            # de-serialized as object for pyarrow < 0.15
            expected = df.assign(a=df.a.astype(object))
>           check_round_trip(df, pa, expected=expected)

pandas/tests/io/test_parquet.py:468: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pandas/tests/io/test_parquet.py:175: in check_round_trip
    compare(repeat)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

repeat = 2

    def compare(repeat):
        for _ in range(repeat):
            df.to_parquet(path, **write_kwargs)
            with catch_warnings(record=True):
                actual = read_parquet(path, **read_kwargs)
    
>           tm.assert_frame_equal(expected, actual, check_names=check_names)
E           AssertionError: Attributes are different
E           
E           Attribute "dtype" are different
E           [left]:  CategoricalDtype(categories=['foo', 'bar', 'baz'], ordered=False)
E           [right]: object

pandas/tests/io/test_parquet.py:171: AssertionError
```
It seems that the `expected` still has the categorical dtype (so the `assign` failed to overwrite the original column), but not sure why this only happens on the mac/windows builds ..

**Comment by jorisvandenbossche** [comment]
Ah, no, I already see ;). In the `expected` for the branch for older pyarrow versions, you still need to convert column `'b'` to object dtype as well

**Comment by galuhsahid** [comment]
> Indeed, I see the same trying your code. That is certainly not expected I think. Can you open an issue for this? (Arrow uses a JIRA tracker: https://issues.apache.org/jira/projects/ARROW/issues; I can also open an issue if you want)

I've opened the issue here: https://issues.apache.org/jira/browse/ARROW-6302

**Comment by jorisvandenbossche** [comment]
Thanks!

**Comment by galuhsahid** [comment]
@jorisvandenbossche I've added the test for the ordered flag. Is there any other test that you think we should have for now?

File: doc/source/user_guide/io.rst [hunk 1]
```diff
@@ -4702,7 +4702,7 @@ Several caveats.
   indexes. This extra column can cause problems for non-Pandas consumers that are not expecting it. You can
   force including or omitting indexes with the ``index`` argument, regardless of the underlying engine.
 * Index level names, if specified, must be strings.
-* Categorical dtypes can be serialized to parquet, but will de-serialize as ``object`` dtype.
+* Categorical dtypes for non-string types can be serialized to parquet, but will de-serialize as ``object`` dtype.
```
[the remark this answers is not in the document]
Comment (galuhsahid) [reply doc/source/user_guide/io.rst]: - I've just tried categorical dtypes using integer and in pyarrow it gets de-serialized into float. Not sure if it's expected, if someone can confirm it'd be great.

```
>>> import pandas as pd
>>> from pandas.io.parquet import read_parquet, to_parquet

>>> df["a"] = pd.Categorical(["a", "b", "c", "a"], categories=["b", "c", "d"], ordered=True)
>>> df["b"] = pd.Categorical([1, 2, 3, 1], categories=[2, 3, 4], ordered=True)

>>> df["a"]
0    NaN
1      b
2      c
3    NaN
Name: a, dtype: category
Categories (3, object): [b < c < d]

>>> df["b"]
0    NaN
1      2
2      3
3    NaN
Name: b, dtype: category
Categories (3, int64): [2 < 3 < 4]

>>> df.to_parquet("test_pyarrow.parquet", engine="pyarrow")
>>> actual_pyarrow = df.read_parquet("test_pyarrow.parquet", engine="fastparquet")

>>> actual_pyarrow["a"]
0    NaN
1      b
2      c
3    NaN
Name: a, dtype: category
Categories (3, object): [b, c, d]

>>> actual_pyarrow["b"]
0    NaN
1    2.0
2    3.0
3    NaN
Name: b, dtype: float64
```

- With fastparquet, both string and non-string types get de-serialized back as category, however it does not preserve order for both, while in pyarrow (for string) we do preserve order. Should we make this clear in the docs? 
```
>>> import pandas as pd
>>> from pandas.io.parquet import read_parquet, to_parquet

>>> df["a"] = pd.Categorical(["a", "b", "c", "a"], categories=["b", "c", "d"], ordered=True)
>>> df["b"] = pd.Categorical([1, 2, 3, 1], categories=[2, 3, 4], ordered=True)

>>> df["a"]
0    NaN
1      b
2      c
3    NaN
Name: a, dtype: category
Categories (3, object): [b < c < d]

>>> df["b"]
0    NaN
1      2
2      3
3    NaN
Name: b, dtype: category
Categories (3, int64): [2 < 3 < 4]

>>> df.to_parquet("test_fastparquet.parquet", engine="fastparquet")
>>> actual_fastparquet = df.read_parquet("test_fastparquet.parquet", engine="fastparquet")

>>> actual_fastparquet["a"]
0    NaN
1      b
2      c
3    NaN
Name: a, dtype: category
Categories (3, object): [b, c, d]

>>> actual_fastparquet["b"]
0    NaN
1      2
2      3
3    NaN
Name: b, dtype: category
Categories (3, int64): [2, 3, 4]
```

**Comment by jorisvandenbossche** [comment]
> I've just tried categorical dtypes using integer and in pyarrow it gets de-serialized into float. Not sure if it's expected, if someone can confirm it'd be great.

For Arrow, that is expected, see https://issues.apache.org/jira/browse/ARROW-6140 (the fact that it is float and not int is because there are missing values)

> With fastparquet, both string and non-string types get de-serialized back as category, however it does not preserve order for both, while in pyarrow (for string) we do preserve order. Should we make this clear in the docs?

That could be mentioned yes.

**Comment by galuhsahid** [comment]
I've updated the doc to explain more about the differences regarding how the categorical support works between pyarrow and fastparquet

File: pandas/tests/io/test_parquet.py [hunk 2]
```diff
@@ -451,11 +453,26 @@ def test_unsupported(self, pa):
...
+
+        # test for ordered flag
+        df["c"] = pd.Categorical(
+            ["a", "b", "c", "a", "c", "b"], categories=["b", "c", "d"], ordered=True
+        )
 
-        # de-serialized as object
-        expected = df.assign(a=df.a.astype(object))
-        check_round_trip(df, pa, expected=expected)
+        if LooseVersion(pyarrow.__version__) >= LooseVersion("0.15.0"):
```
Comment (TomAugspurger) [inline pandas/tests/io/test_parquet.py]: @jorisvandenbossche this isn't released yet, right? Should we wait to merge until 0.15.0 is released?

Comment (jorisvandenbossche) [reply pandas/tests/io/test_parquet.py]: Yes, this isn't released yet. I can assure that it runs locally for me on Arrow master (if I change the version check to `> LooseVersion("0.14.1.dev")`), so I am OK to merge this, but also fine to wait a bit more (0.15.0 will normally happen somewhere end of next week)

**Review by jorisvandenbossche — APPROVED** [review]

File: pandas/tests/io/test_parquet.py (hunk 2 above)
Comment (galuhsahid) [reply pandas/tests/io/test_parquet.py]: Since Arrow 0.15.0 has been released, can we merge this now?

File: doc/source/whatsnew/v1.0.0.rst [hunk 3]
```diff
@@ -174,6 +174,7 @@ Categorical
 - Added test to assert the :func:`fillna` raises the correct ValueError message when the value isn't a value from categories (:issue:`13628`)
 - Bug in :meth:`Categorical.astype` where ``NaN`` values were handled incorrectly when casting to int (:issue:`28406`)
 - :meth:`Categorical.searchsorted` and :meth:`CategoricalIndex.searchsorted` now work on unordered categoricals also (:issue:`21667`)
+- Added test to assert roundtripping to parquet with :func:`DataFrame.to_parquet` or :func:`read_parquet` will preserve Categorical dtypes for string types (:issue:`27955`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: this note is not really needed, but ok

File: doc/source/user_guide/io.rst [hunk 4]
```diff
@@ -4734,7 +4735,8 @@ See the documentation for `pyarrow <https://arrow.apache.org/docs/python/>`__ an
                       'd': np.arange(4.0, 7.0, dtype='float64'),
                       'e': [True, False, True],
                       'f': pd.date_range('20130101', periods=3),
-                      'g': pd.date_range('20130101', periods=3, tz='US/Eastern')})
+                      'g': pd.date_range('20130101', periods=3, tz='US/Eastern'),
+                      'h': pd.Categorical(list('abc'))})
```
Comment (jreback) [inline doc/source/user_guide/io.rst]: can we add an ordered categorical as well?

**Review by jreback — CHANGES_REQUESTED** [review]
pls merge master

**Comment by jorisvandenbossche** [comment]
@galuhsahid Thanks a lot for this PR! (and for the fixes on the pyarrow side as well! :-))

**Pull request merged**